pkg/controller/csi_handler.go

```diff
@@ -93,6 +101,99 @@ func (h *csiHandler) Init(vaQueue workqueue.RateLimitingInterface, pvQueue workq
 	h.pvQueue = pvQueue
 }
 
+// ReconcileVA lists volumes from the CSI Driver and reconciles the attachment
+// status with the corresponding VolumeAttachment object. If the attachment
+// status of the volume is different from the state on the VolumeAttachment the
+// VolumeAttachment object is patched to the correct state.
+func (h *csiHandler) ReconcileVA() error {
+	ctx, cancel := context.WithTimeout(context.Background(), h.timeout)
+	defer cancel()
+
+	// Loop over all volume attachment objects
+	vas, err := h.vaLister.List(labels.Everything())
+	if err != nil {
+		return errors.New("failed to list all VolumeAttachment objects")
+	}
+
+	// List volumes after getting volume attachment objects to prevent race
+	// condition where actual state is changed between ListVolumes and
+	// VolumeAttachment List. When listing VolumeAttachments first we can
+	// guarantee any race conditions will result in API object UPDATE failures
+	// instead of stomping on the VolumeAttachment with outdated data
+	published, err := h.volumeLister.ListVolumes(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to ListVolumes: %v", err)
+	}
+
+	for _, va := range vas {
+		nodeID, ok := va.Annotations[vaNodeIDAnnotation]
+		if !ok {
+			klog.Warning("failed to find node ID in VolumeAttachment annotation")
+			continue
+		}
+		pvSpec, err := h.getPVSpec(va)
+		if err != nil {
+			klog.Warningf("failed to get PV Spec: %v", err)
+			continue
+		}
+
+		source, err := getCSISource(pvSpec)
+		if err != nil {
+			klog.Warningf("failed to get CSI Source: %v", err)
+			continue
+		}
+
+		volumeHandle, _, err := GetVolumeHandle(source)
+		if err != nil {
+			klog.Warningf("failed to get volume handle %v", err)
+			continue
+		}
+		attachedStatus := va.Status.Attached
+
+		volumeHandle, err = h.translator.RepairVolumeHandle(source.Driver, volumeHandle, nodeID)
+		if err != nil {
+			klog.Warningf("failed to repair volume handle %s for driver %s: %v", volumeHandle, source.Driver, err)
+			continue
+		}
+
+		// If ListVolumes Attached Status is different, update.
```

Comment:
Maybe update this comment to, "Check whether the volume is published to this node"